katz/src/main/kotlin/katz/Either.kt

```diff
@@ -0,0 +1,194 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package katz
+
+/**
+ * Port of https://github.com/scala/scala/blob/v2.12.1/src/library/scala/util/Either.scala
+ *
+ * Represents a value of one of two possible types (a disjoint union.)
+ * An instance of Either is either an instance of [Left] or [Right].
+ */
+sealed class Either<out A : Any, out B : Any> {
+
+    /**
+     * Returns `true` if this is a [Right], `false` otherwise.
+     *
+     * Example:
+     * ```
+     * Left("tulip").isRight           // Result: false
+     * Right("venus fly-trap").isRight // Result: true
+     * ```
+     */
+    abstract val isRight: Boolean
+
+    /**
+     * Returns `true` if this is a [Left], `false` otherwise.
+     *
+     * Example:
+     * ```
+     * Left("tulip").isLeft           // Result: true
+     * Right("venus fly-trap").isLeft // Result: false
+     * ```
+     */
+    abstract val isLeft: Boolean
+
+    /**
+     * Applies `fa` if this is a [Left] or `fb` if this is a [Right].
+     *
+     * Example:
+     * ```
+     * val result: Either<Exception, Value> = possiblyFailingOperation()
+     * result.fold(
+     *      { log("operation failed with $it") },
+     *      { log("operation succeeded with $it") }
+     * )
+     * ```
+     *
+     * @param fa the function to apply if this is a [Left]
+     * @param fb the function to apply if this is a [Right]
+     * @return the results of applying the function
+     */
+    inline fun <C> fold(fa: (A) -> C, fb: (B) -> C): C = when (this) {
+        is Right -> fb(b)
+        is Left -> fa(a)
+    }
+
+    /**
+     * If this is a `Left`, then return the left value in `Right` or vice versa.
+     *
+     * Example:
+     * ```
+     * Left("left").swap()   // Result: Right("left")
+     * Right("right").swap() // Result: Left("right")
+     * ```
+     */
+    fun swap(): Either<B, A> =
+            fold({ Right(it) }, { Left(it) })
+
+    /**
+     * The given function is applied if this is a `Right`.
+     *
+     * Example:
+     * ```
+     * Right(12).map { "flower" } // Result: Right("flower")
+     * Left(12).map { "flower" }  // Result: Left(12)
+     * ```
+     */
+    inline fun <C : Any> map(f: (B) -> C): Either<A, C> =
+            fold({ Left(it) }, { Right(f(it)) })
+
+    /**
+     * Returns `true` if this is a [Right] and its value is equal to `elem` (as determined by `==`),
+     * returns `false` otherwise.
+     *
+     * Example:
+     * ```
+     * Right("something").contains { "something" } // Result: true
+     * Right("something").contains { "anything" }  // Result: false
+     * Left("something").contains { "something" }  // Result: false
+     *  ```
+     *
+     * @param elem    the element to test.
+     * @return `true` if the option has an element that is equal (as determined by `==`) to `elem`, `false` otherwise.
+     */
+    fun <BB, B> contains(elem: () -> BB): Boolean where B : BB =
```

Comment:
I wanted to make the value lazy in here, since if this Either is a Left, then `elem` doesn't have to be evaluated, in case is a call to a method or something similar. What do you think?